Assets/Source/Components/NavigationMesh/NavigationMeshComponent.cs

```diff
@@ -0,0 +1,135 @@
+        // Generate the initial nav mesh
+        private void GenerateNavigationMesh()
+        {
+            for (var ix = 0; ix < gridWidth; ix++)
+            {
+                nodes[ix] = new Node[gridHeight];
+                for (var iy = 0; iy < gridHeight; iy++)
+                {
+                    var center = new Vector2(x: (ix * tileSize) + transform.position.x, y: (iy * tileSize) + transform.position.y);
+                    var isSolid = TileContainsSolid(center);
+
+                    nodes[ix][iy] = new Node(center, ix, iy, isSolid);
+                }
+            }
+        }
+
+        // Returns true if any solid exists in a radius of 'tileSize' around the specified center position
+        private bool TileContainsSolid(Vector2 center) => Physics2D.OverlapArea(new Vector2(center.x - (tileSize / 2), center.y - (tileSize / 2)),
+                                                                                new Vector2(center.x + (tileSize / 2), center.y + (tileSize / 2))) != null;
+
+        
+
+        public override void ComponentUpdate()
+        {
+            if (syncEveryFrame)
+            {
+                UpdateTiles();
+            }
+            base.ComponentUpdate();
+        }
+
+        /// <summary>
+        /// Used to clone the 2d array of nodes so we can modify them safely
+        /// </summary>
+        /// <returns></returns>
+        public Node[][] CloneGrid() => nodes.Select(n => n.ToArray()).ToArray();
+
+
+        // For each tile, update its "isSolid" status.
+        private void UpdateTiles()
+        {
+            if (nodes != null)
+            {
+                foreach (var nodeRow in nodes)
+                {
+                    foreach (var node in nodeRow)
+                    {
+                        node.IsSolid = TileContainsSolid(node.Center);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Returns a tuple which contains the position of the nearest node.  If the specified position is beyond the boundaries 
+        /// of the navigation mesh, it will return the closest node it can get within the confines of the grid. 
+        /// </summary>
+        /// <param name="worldPosition">Position in world space </param>
+        /// <returns>x / y position of the node</returns>
+        public (int ix, int iy) FindNearestNodeIndex(Vector2 worldPosition)
+        {
+            // Basically reverse the function to plot the grid square to figure out its index
+            var row = (int)Mathf.Clamp(Mathf.Round((worldPosition.x - transform.position.x) / tileSize), 0, gridWidth - 1);
+            var col = (int)Mathf.Clamp(Mathf.Round((worldPosition.y - transform.position.y) / tileSize), 0, gridHeight - 1);
+            return (row, col);
+        }
+
+
+        // todo: change to this line because otherwise that grid will get annoying 
+        //private void OnDrawGizmosSelected()
+        private void OnDrawGizmos()
```

Comment:
These gizmos are awesome for debugging and during development, should we make a ticket to clean these up one the game nears completion?

Reply:
I don't think so, the gizmos don't have any affect on the final build.